api/main.py

```diff
@@ -138,6 +139,41 @@ async def post_user(
     return obj
 
 
+@app.post('/group/{group_name}', response_model=UserGroup)
```

Comment:
One thing, maybe we don't need `group_name` in the URL?  This could be sent as part of the POST data.  I think having the group name in POST and a group ID in GET is not ideal as it creates some asymmetry between the query methods.

Reply:
Since the `group_name` is a string and not a Pydantic model, by default, it will be treated as a path or query parameter.
See https://fastapi.tiangolo.com/tutorial/body/#request-body-path-query-parameters
We need to create a model if we want to receive it in the request body. Or we can use `Body(...)` as suggested here https://github.com/tiangolo/fastapi/discussions/9010. But this way we won't have any validation for the input.
Also, I don't see much of an issue here as we have two separate methods i.e. GET and POST.

Reply:
I mean this, just like when creating a new node:
```python
@app.post('/group', response_model=UserGroup)
async def post_group(group: UserGroup, current_user: User = Security(get_user, scopes=["admin"])):
```

Reply:
About the URI paths, the only inconsistency I can think of is that we don't know the id before creating a group (same with users and nodes...).  Apart from this, we can have:
```
http://baseurl.com/group/<group-id>
```
and then use POST, PUT and GET with the same URI path to manage this resource.

Note: In an ideal world we could let the client create the IDs (like with KCIDB) but I think this would be suboptimal from a MongoDB point of view as we can't create ObjectID() with any arbitrary ID.  So we would need to create another index and have both the MongoDB `_id` and the user-supplied ID.

Reply:
> About the URI paths, the only inconsistency I can think of is that we don't know the id before creating a group (same with users and nodes...). Apart from this, we can have:
> 
> ```
> http://baseurl.com/group/<group-id>
> ```

No, we don't need an id to create a user group. Please look at the sample command below:
```
$ curl -X 'POST' 'http://localhost:8001/latest/group' -H 'accept: application/json'   -H 'Content-Type: application/json'  -H 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJhZG1pbiIsInNjb3BlcyI6WyJhZG1pbiJdfQ.4fGx_hBFdD-1JG7z38A-E0v8vkjnCa3Zj6a1p1fqWyk' -d '{"name": "kernelci"}'
{"_id":"648ff894bd39930355ed16ad","name":"kernelci"}
```
And yes, we already have `groups` and `group/{group_id}` endpoints implemented.

Reply:
Yes I know we don't need an ID, and that's precisely the only inconsistency I can see with how the URIs are designed.  Ideally, the same URI should be used for creating, updating, getting, deleting (all the HTTP verbs) a same resource.  I think we can have an exception here for creating objects as it's a pretty common use-case.  If we need the API to be closer to KCIDB then we might have to revisit this and rely on the client to create an ID.

Reply:
In what scenarios we would need to let the client create an ID?

Reply:
KCIDB is based on the design principles that the same data sent to the API should be retrievable, so that means all the fields have to be provided by the user.  Also it's designed for gathering data from multiple CI systems that already have their own database with their own IDs, like KernelCI is sending data with the IDs generated by MongoDB to KCIDB.  So if we wanted to align the new KernelCI API with KCIDB's schema we would also require users to provide IDs, and probably also timestamps etc.  But it's not a requirement for now, I think we can consider this as a long-term potential goal.  We should just try to avoid designing the new API in ways that will make it harder to align with KCIDB later.

Reply:
So in that case we would need to introduce another `id` field that the user can provide as you suggested.
I think we can do that easily without messing with the existing schema.